### DISCUSSION TITLE
How can I extract embeddings?

### DISCUSSION THREAD

**Discussion by isaacOnline** [openai/whisper#85] [Q&A] [answered]
Hi! Would someone be able to provide an example showing how to extract embeddings?

**Comment by isaacOnline** [comment, marked as answer]
I created a fork to extract embeddings [here](https://github.com/isaacOnline/whisper/tree/extract-embeddings). The embeddings get stored as a part of the transcription result that's returned when you call `model.transcribe()` and are done at the segment level.

**Reply by mallorbc** [reply]
What is the shape of the embeddings?  Why do you want them in the first place?

**Reply by haiderasad** [reply]
hi @isaacOnline got the embeddings but only works with cpu, on cuda its giving error , any idea?

**Reply by isaacOnline** [reply]
> What is the shape of the embeddings? Why do you want them in the first place?

I’m extracting embeddings for each segment that the models splits the audio into, and for each segment I’m extracting one set of embeddings for the encoder and one set for the decoder. Each of those sets will have four dimensions: 
* One corresponding to the number of samples you're processing
* One corresponding to the number of transformer layers in the Whisper model that you’re using
* One corresponding to the the length of the segment
* One corresponding to the width of the Whisper model you’re using

I'm using them for adapting Whisper to another task.

**Reply by isaacOnline** [reply]
> hi @isaacOnline got the embeddings but only works with cpu, on cuda its giving error , any idea?

Hmm, it’s been working fine for me on cuda. Are you sure you have enough room on your GPU for the model version you’re using? Other than that I can’t provide much help without a reproducible example.

**Reply by haiderasad** [reply]
> > What is the shape of the embeddings? Why do you want them in the first place?
> 
> I’m extracting embeddings for each segment that the models splits the audio into, and for each segment I’m extracting one set of embeddings for the encoder and one set for the decoder. Each of those sets will have four dimensions:
> 
> * One corresponding to the number of samples you're processing
> * One corresponding to the number of transformer layers in the Whisper model that you’re using
> * One corresponding to the the length of the segment
> * One corresponding to the width of the Whisper model you’re using
> 
> I'm using them for adapting Whisper to another task.

can you tell what the decoder embeddings represent overall? the finalized text in the specified language or is it related to speech embeddings

**Reply by isaacOnline** [reply]
Unfortunately I've mostly been using the encoder embeddings myself.

**Reply by tusharagarwal25** [reply]
hi @haiderasad, were you able to work with the embeddings?

**Reply by tusharagarwal25** [reply]
hi @isaacOnline, getting this error: UnboundLocalError: local variable 'encoder_embeddings' referenced before assignment

Can you please help here?
<img width="1275" alt="Screenshot 2023-05-03 at 12 28 25 PM" src="https://user-images.githubusercontent.com/12660586/235850361-7f462277-da73-4e36-bfe8-53e1925d6a1a.png">

**Reply by isaacOnline** [reply]
The stuff I wrote is in a separate  "extract-embeddings" branch that you'll need to checkout. I think I've gotten the error you show when I've forgotten to switch to the right branch. You may also need to do the pip install:

git checkout extract-embeddings
pip install --editable ./

**Comment by outliyr** [comment]
Did you ever try using the embeddings for speaker diarization, and if not what were you using the embedding for?

**Comment by saeidmokaram** [comment]
Thanks @isaacOnline, This is very helpful. I was wondering what sort of information we should expect to be captured by the encoder_embeddings and what information by the decoder_embeddings?

**Comment by alxmares** [comment]
You can extract embeddings from the Whisper model using the Transformers library as follows:

```
processor = WhisperProcessor.from_pretrained("openai/whisper-base")
model = WhisperModel.from_pretrained("openai/whisper-base")

inputs = processor(speech_array, sampling_rate=16000, return_tensors="pt")

with torch.no_grad():
    encoder_outputs = model.encoder(inputs.input_features)

embeddings = encoder_outputs.last_hidden_state
```
